pkg/config/expand.go

```diff
@@ -247,60 +246,60 @@ func (dc *DeploymentConfig) combineLabels() error {
-	if _, ok := mod.WrapSettingsWith[labels]; ok {
-		return nil // Do nothing
+	if !moduleHasInput(*mod, labels) {
+		return nil // no op
 	}
 
-	// Check if labels are set for this module
-	if !moduleHasInput(*mod, labels) {
-		return nil
+	cur := mod.Settings.Get(labels)
+	extra := map[string]cty.Value{roleLabel: cty.StringVal(getRole(mod.Source))}
+
+	if mod.Kind == TerraformKind {
+		mod.Settings.Set(labels, mergeLabelsTf(extra, cur))
+	} else if mod.Kind == PackerKind {
+		gl := dc.Config.Vars.Get(labels).AsValueMap()
+		merged, err := mergeLabelsPkr(gl, extra, cur)
+		if err != nil {
+			return err
+		}
+		mod.Settings.Set(labels, merged)
 	}
+	return nil
+}
 
+// Terraform labels are `merge(var.labels, {ghpc_role="foo"}, [module labels])`
+func mergeLabelsTf(extra map[string]cty.Value, cur cty.Value) cty.Value {
+	args := []cty.Value{
+		GlobalRef("labels").AsExpression().AsValue(),
+		cty.ObjectVal(extra),
+	}
+	if !cur.IsNull() {
+		args = append(args, cur)
+	}
+	return FunctionCallExpression("merge", args...).AsValue()
+}
+
+// Packer doesn't support `merge`, so merge it here.
+func mergeLabelsPkr(global map[string]cty.Value, extra map[string]cty.Value, cur cty.Value) (cty.Value, error) {
 	modLabels := map[string]cty.Value{}
-	if mod.Settings.Has(labels) {
-		// Cast into map so we can index into them
-		v := mod.Settings.Get(labels)
-		ty := v.Type()
+	if !cur.IsNull() {
+		ty := cur.Type()
 		if !ty.IsObjectType() && !ty.IsMapType() {
-			return fmt.Errorf("%s, Module %s, labels type: %s",
-				errorMessages["settingsLabelType"], mod.ID, ty.FriendlyName())
+			return cty.NilVal, fmt.Errorf("%s,labels type: %s", errorMessages["settingsLabelType"], ty.FriendlyName())
 		}
-		if v.AsValueMap() != nil {
-			modLabels = v.AsValueMap()
+		if cur.AsValueMap() != nil {
+			modLabels = cur.AsValueMap()
 		}
 	}
-	// Add the role (e.g. compute, network, etc)
-	if _, exists := modLabels[roleLabel]; !exists {
-		modLabels[roleLabel] = cty.StringVal(getRole(mod.Source))
-	}
-
-	if mod.Kind == TerraformKind {
-		// Terraform module labels to be expressed as
-		// `merge(var.labels, { ghpc_role=..., **settings.labels })`
-		mod.WrapSettingsWith[labels] = []string{"merge(", ")"}
-		ref := GlobalRef(labels).AsExpression()
-		args := []cty.Value{ref.AsValue(), cty.ObjectVal(modLabels)}
-		mod.Settings.Set(labels, cty.TupleVal(args))
-	} else if mod.Kind == PackerKind {
-		g := dc.Config.Vars.Get(labels).AsValueMap()
-		mod.Settings.Set(labels, cty.ObjectVal(mergeLabels(modLabels, g)))
-	}
-	return nil
+	return cty.ObjectVal(mergeMaps(global, extra, modLabels)), nil
 }
 
-// mergeLabels returns a new map with the keys from both maps. If a key exists in both maps,
-// the value from the first map is used.
-func mergeLabels(a map[string]cty.Value, b map[string]cty.Value) map[string]cty.Value {
+// see https://developer.hashicorp.com/terraform/language/functions/merge
+func mergeMaps(ms ...map[string]cty.Value) map[string]cty.Value {
```

Comment:
Please re-create the part of the comment that explicitly states which elements win in case of conflict (i.e. first, last)